### PR TITLE
[kenlm] Update port

### PR DESCRIPTION
**Pull request by jacobkahn**
Updates the KenLM port. I've added a `kenlmConfig.cmake` in https://github.com/kpu/kenlm/pull/334 and https://github.com/kpu/kenlm/pull/339. Updating the port to contain those so we can use modern CMake targets with the port. This generally fixes a bunch of problems with downstream projects that depend on kenlm and removes the need to specify a bunch of env vars just to find relevant bits.

Still depends on https://github.com/kpu/kenlm/pull/342.

A fix to some const things for msvc has also been merged upstream, so remove that patch.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

Same triplets as before. No changes.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  

Yes

cc @kpu, @tlikhomanenko, @xuqiantong, @vineelpratap, @padentomasello, @andresy, @syhw

### CODE REVIEW

**Comment by jacobkahn** [comment]
Seems like Windows builds are failing – I have no means to debug - @kpu can you help here/can someone share logs from CI?

**Comment by JonLiu1993** [comment]
@jacobkahn ,Thanks for your pr,
It looks like x86-Windows, x64-windows, x64-windows-
static reported the same error,could you take a look:
````
[1/11] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.20.2-windows\cmake-3.20.2-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=lm\CMakeFiles\query.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\x64\link.exe  lm\CMakeFiles\query.dir\query_main.cc.obj  /out:bin\query.exe /implib:bin\query.lib /pdb:bin\query.pdb /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL /subsystem:console  lib\kenlm.lib  lib\kenlm_util.lib  D:\installed\x64-windows\debug\lib\zlibd.lib  D:\installed\x64-windows\debug\lib\bz2d.lib  D:\installed\x64-windows\debug\lib\lzmad.lib  optimized.lib  D:\installed\x64-windows\lib\boost_program_options-vc140-mt.lib  debug.lib  D:\installed\x64-windows\debug\lib\boost_program_options-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_system-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_system-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_thread-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_thread-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_unit_test_framework-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_unit_test_framework-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_chrono-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_chrono-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_date_time-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_date_time-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_atomic-vc140-mt.lib  optimized.lib  D:\installed\x64-windows\lib\boost_program_options-vc140-mt.lib  debug.lib  D:\installed\x64-windows\debug\lib\boost_program_options-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_system-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_system-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_thread-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_thread-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_unit_test_framework-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_unit_test_framework-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_chrono-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_chrono-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_date_time-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_date_time-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_atomic-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_atomic-vc140-mt-gd.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
FAILED: bin/query.exe 
cmd.exe /C "cd . && D:\downloads\tools\cmake-3.20.2-windows\cmake-3.20.2-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=lm\CMakeFiles\query.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\x64\link.exe  lm\CMakeFiles\query.dir\query_main.cc.obj  /out:bin\query.exe /implib:bin\query.lib /pdb:bin\query.pdb /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL /subsystem:console  lib\kenlm.lib  lib\kenlm_util.lib  D:\installed\x64-windows\debug\lib\zlibd.lib  D:\installed\x64-windows\debug\lib\bz2d.lib  D:\installed\x64-windows\debug\lib\lzmad.lib  optimized.lib  D:\installed\x64-windows\lib\boost_program_options-vc140-mt.lib  debug.lib  D:\installed\x64-windows\debug\lib\boost_program_options-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_system-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_system-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_thread-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_thread-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_unit_test_framework-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_unit_test_framework-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_chrono-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_chrono-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_date_time-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_date_time-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_atomic-vc140-mt.lib  optimized.lib  D:\installed\x64-windows\lib\boost_program_options-vc140-mt.lib  debug.lib  D:\installed\x64-windows\debug\lib\boost_program_options-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_system-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_system-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_thread-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_thread-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_unit_test_framework-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_unit_test_framework-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_chrono-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_chrono-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_date_time-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_date_time-vc140-mt-gd.lib  D:\installed\x64-windows\lib\boost_atomic-vc140-mt.lib  D:\installed\x64-windows\debug\lib\boost_atomic-vc140-mt-gd.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\x64\link.exe lm\CMakeFiles\query.dir\query_main.cc.obj /out:bin\query.exe /implib:bin\query.lib /pdb:bin\query.pdb /version:0.0 /machine:x64 /nologo /debug /INCREMENTAL /subsystem:console lib\kenlm.lib lib\kenlm_util.lib D:\installed\x64-windows\debug\lib\zlibd.lib D:\installed\x64-windows\debug\lib\bz2d.lib D:\installed\x64-windows\debug\lib\lzmad.lib optimized.lib D:\installed\x64-windows\lib\boost_program_options-vc140-mt.lib debug.lib D:\installed\x64-windows\debug\lib\boost_program_options-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_system-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_system-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_thread-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_thread-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_unit_test_framework-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_unit_test_framework-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_chrono-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_chrono-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_date_time-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_date_time-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_atomic-vc140-mt.lib optimized.lib D:\installed\x64-windows\lib\boost_program_options-vc140-mt.lib debug.lib D:\installed\x64-windows\debug\lib\boost_program_options-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_system-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_system-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_thread-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_thread-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_unit_test_framework-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_unit_test_framework-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_chrono-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_chrono-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_date_time-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_date_time-vc140-mt-gd.lib D:\installed\x64-windows\lib\boost_atomic-vc140-mt.lib D:\installed\x64-windows\debug\lib\boost_atomic-vc140-mt-gd.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:lm\CMakeFiles\query.dir/intermediate.manifest lm\CMakeFiles\query.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'optimized.lib'
ninja: build stopped: subcommand failed.
````

File: ports/kenlm/portfile.cmake [hunk 1]
```diff
@@ -3,17 +3,17 @@ vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
...
     OUT_SOURCE_PATH SOURCE_PATH
     REPO kpu/kenlm
-    REF 1f054617eca14eae921e987b4b4eeb2b1d91de6b
-    SHA512 c18f9c22fbbb1f54ebe9c3b771fb2d7c09d502141d1b3645cff9db44cc51b3c976311ff0db79b60f410622579d043f185c56a4c7386e1b0ba8708e433238968b
+    REF bdf3c71a34a874de11ab02f23ebe0a0b877c27ef
+    SHA512 3782cf4b08b5686ea320fa248012c780faa0417fa5b23f2645bc8c92f7741ec8b8c7f81cd3f58676b1d1ea5817d1b56eaffa608ea993e3d213b63fcba44e2afb
     HEAD_REF master
     PATCHES 
         fix-boost.patch
-        fix-const-overloaded.patch
```
Comment (JonLiu1993) [inline ports/kenlm/portfile.cmake:11]: Why can't I see the deleted patch in the modification record in pr

File: ports/kenlm/portfile.cmake [hunk 2]
```diff
@@ -3,17 +3,17 @@ vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
...
 )
 
 file(REMOVE ${SOURCE_PATH}/cmake/modules/FindEigen3.cmake)
 
 vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
     interpolate ENABLE_INTERPOLATE
```
Comment (JonLiu1993) [inline ports/kenlm/portfile.cmake:17]: There should be an indentation of four spaces relative to'FEATURES'

**Comment by JonLiu1993** [comment]
@jacobkahn ,Could you please take a look?

**Comment by JonLiu1993** [comment]
Ping @jacobkahn  for response, Is work still being done for this PR?

**Comment by jacobkahn** [comment]
@kpu do you have any insight on this Windows error? I can't tell if it's build related or compiler-related.

**Comment by kpu** [comment]
Boost linkage on windows is a mystery to me.  I tried and failed to fix it in https://github.com/kpu/kenlm/tree/semaphore (which despite the name ended up being a windows experimentation branch).

**Comment by jacobkahn** [comment]
@kpu it might make sense to just remove Windows support from the vcpkg port for the time being if that's the case — thoughts?

**Comment by JonLiu1993** [comment]
Pinging @jacobkahn  for response. Is work still being done for this PR?

**Comment by JonLiu1993** [comment]
@jacobkahn for response. Is work still being done for this PR?

**Comment by JonLiu1993** [comment]
@jacobkahn ,Is this PR being worked on?

**Comment by jacobkahn** [comment]
@JonLiu1993 — I'll be returning to this PR in about 3 weeks. You can close it for now if that's easier and I can reopen it.

**Comment by JonLiu1993** [comment]
@jacobkahn ,I close this pr, if you need to fix it, please reopen it

**Pull request closed without merging**